### PR TITLE
Parse M.2s, propagate parent information about zpools to Nexus

### PR DESCRIPTION
**Pull request by smklein**
## Sled Agent (and `illumos-utils`, `sled-hardware`)

- Parse Zpools with the `oxi_` prefix, in addition to those with the `oxp_` prefix
  - This is a necessary follow-up to https://github.com/oxidecomputer/facade/issues/11
- Pass information about the parent disk up to Nexus when notifying about zpools.
  - This is necessary for Nexus to identify "what should be provisioned to this zpool", by correlating the pool with the corresponding physical disk.

## Nexus

- Update the "Zpool PUT" API to include the identifying information for the parent disk
- As part of the implementation of Zpool PUT, look up the parent disk, and insert that parent disk UUID into the Zpool DB row.

This is the first part of https://github.com/oxidecomputer/omicron/issues/2688

### CODE REVIEW

File: nexus/types/src/internal_api/params.rs [hunk 1]
```diff
@@ -94,6 +94,11 @@ pub struct PhysicalDiskDeleteRequest {
 pub struct ZpoolPutRequest {
     /// Total size of the pool.
     pub size: ByteCount,
+
+    // Information to identify the disk to which this zpool belongs
+    pub vendor: String,
```
Comment (davepacheco) [inline nexus/types/src/internal_api/params.rs:99]: nit: prefix here, like "disk_vendor"?  (it seems a little confusing to call these the "vendor", "serial", and "model" _of the zpool_)

File: sled-agent/src/storage_manager.rs [hunk 2]
```diff
@@ -129,28 +129,23 @@ pub enum Error {
...
-        name: &ZpoolName,
-        disk_path: Option<PathBuf>,
-    ) -> Result<Pool, Error> {
+    fn new(name: ZpoolName, parent: DiskIdentity) -> Result<Pool, Error> {
         let info = Zpool::get_info(&name.to_string())?;
 
         // NOTE: This relies on the name being a UUID exactly.
         // We could be more flexible...
-        Ok(Pool { id: name.id(), info, zones: HashMap::new(), disk_path })
+        Ok(Pool { name, info, zones: HashMap::new(), parent })
```
Comment (davepacheco) [inline sled-agent/src/storage_manager.rs:148]: The comment above this line seems not quite true?

**Review by davepacheco — APPROVED** [review]

File: nexus/db-queries/src/db/datastore/physical_disk.rs [hunk 3]
```diff
@@ -90,6 +91,36 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
+    pub async fn physical_disk_lookup(
```
[the remark this answers is not in the document]
Comment (smklein) [reply nexus/db-queries/src/db/datastore/physical_disk.rs:94]: I believe we already do: https://github.com/oxidecomputer/omicron/blob/e4a5dd0029946c0f7f56f4858475c2935e0eb9a9/nexus/db-queries/src/db/lookup.rs#L678-L685

The primary key is already an ID, but I'm not sure how to say "this object can also be uniquely looked-up by vendor/serial/model".

File: nexus/db-queries/src/db/datastore/physical_disk.rs [hunk 4]
```diff
@@ -90,6 +91,36 @@ impl DataStore {
...
+        model: String,
+    ) -> LookupResult<db::model::PhysicalDisk> {
+        opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
+        use db::schema::physical_disk::dsl;
+        let disk = dsl::physical_disk
+            .filter(dsl::vendor.eq(vendor.clone()))
+            .filter(dsl::serial.eq(serial.clone()))
+            .filter(dsl::model.eq(model.clone()))
+            .select(PhysicalDisk::as_select())
+            .first_async(self.pool())
```
[the remark this answers is not in the document]
Comment (smklein) [reply nexus/db-queries/src/db/datastore/physical_disk.rs]: Updated

File: nexus/types/src/internal_api/params.rs (hunk 1 above)
Comment (smklein) [reply nexus/types/src/internal_api/params.rs]: Updated for vendor, serial, and model.

File: sled-agent/src/storage_manager.rs (hunk 2 above)
Comment (smklein) [reply sled-agent/src/storage_manager.rs:145]: Comment removed

File: nexus/db-queries/src/db/datastore/physical_disk.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (smklein) [reply nexus/db-queries/src/db/datastore/physical_disk.rs]: Went ahead and updated the PK columns here: 1563a70d69ee02015495405b62d164e24bbe2cc9 -- seems to work

[the remark this answers is not in the document]
Comment (smklein) [reply nexus/db-queries/src/db/datastore/physical_disk.rs]: Okay, actually, this required some follow-up changes in https://github.com/oxidecomputer/omicron/pull/2708/commits/d997cc31720ab4db377e41dd510994e11105235ctoo

**Pull request merged**